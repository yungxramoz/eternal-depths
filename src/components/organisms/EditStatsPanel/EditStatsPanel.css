.attribute {
  display: flex !important;
  flex-direction: row;
  justify-content: space-between;
}

.attr-controller {
  display: flex;
  flex-direction: row;
  align-items: center;
  column-gap: 1rem;
}
