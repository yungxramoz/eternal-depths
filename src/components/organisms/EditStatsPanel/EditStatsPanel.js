import React, {useState} from 'react'
import IconButton from '../../molecules/IconButton/IconButton'
import RpgContainer from '../../templates/RpgContainer/RpgContainer'
import RpgSeparator from '../../templates/RpgSeperator/RpgSeparator'
import './EditStatsPanel.css'

const EditStatsPanel = ({ localStats, initialPoints, assignedPoints, setAssignedPoints }) => {
  const [points, setPoints] = useState(initialPoints)

  const updateStats = (stat, value) => {
    if (value < 0 && assignedPoints[stat] === 0) return
    if (value > 0 && points === 0) return
    setAssignedPoints({
      ...assignedPoints,
      [stat]: assignedPoints[stat] + value,
    })
    setPoints(points - value)
  }

  const isMinusDisabled = (stat) => {
    return assignedPoints[stat] === 0
  }

  const isPlusDisabled = () => {
    return points === 0
  }

  return (
    <RpgContainer golden className="attribute-container">
      <h2>Attributes</h2>
      <p>Available points: {points}</p>
      <RpgSeparator golden={false} />
      {Object.keys(localStats).map((stat) => (
        <div className="attribute" key={stat}>
          <p>{stat.charAt(0).toUpperCase() + stat.slice(1)}</p>
          <div className="attr-controller">
            <IconButton
              icon="minus"
              size="medium"
              onClick={() => updateStats(stat, -1)}
              disabled={isMinusDisabled(stat)}
            />
            <p>{localStats[stat] + assignedPoints[stat]}</p>
            <IconButton
              icon="plus"
              size="medium"
              onClick={() => updateStats(stat, 1)}
              disabled={isPlusDisabled()}
            />
          </div>
        </div>
      ))}
    </RpgContainer>
  )
}

export default EditStatsPanel