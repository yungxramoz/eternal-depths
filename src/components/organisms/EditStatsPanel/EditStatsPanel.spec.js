import React from 'react'
import TestRenderer from 'react-test-renderer'
import EditStatsPanel from './EditStatsPanel'

describe('EditStatsPanel', () => {
  const localStats = { strength: 0, agility: 0 }
  const initialPoints = 10
  const assignedPoints = { strength: 0, agility: 0 }
  const setAssignedPoints = jest.fn()

  it('renders EditStatsPanel component', () => {
    const testRenderer = TestRenderer.create(
      <EditStatsPanel
        localStats={localStats}
        initialPoints={initialPoints}
        assignedPoints={assignedPoints}
        setAssignedPoints={setAssignedPoints}
      />
    )
    const testInstance = testRenderer.root

    expect(testInstance.findByType('h2').children).toEqual(['Attributes'])
  })

  it('renders the correct number of available points', () => {
    const testRenderer = TestRenderer.create(
      <EditStatsPanel
        localStats={localStats}
        initialPoints={initialPoints}
        assignedPoints={assignedPoints}
        setAssignedPoints={setAssignedPoints}
      />
    )
    const testInstance = testRenderer.root

    expect(testInstance.findAllByType('p')[0].children).toEqual(['Available points: ', '10'])
  })

  it('renders the correct number of attributes', () => {
    const tree = TestRenderer.create(
      <EditStatsPanel
        localStats={localStats}
        initialPoints={initialPoints}
        assignedPoints={assignedPoints}
        setAssignedPoints={setAssignedPoints}
      />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders the correct attribute names', () => {
    const testRenderer = TestRenderer.create(
      <EditStatsPanel
        localStats={localStats}
        initialPoints={initialPoints}
        assignedPoints={assignedPoints}
        setAssignedPoints={setAssignedPoints}
      />
    )
    const testInstance = testRenderer.root

    expect(testInstance.findAllByProps({ className: 'attribute' })[0].children[0].children).toEqual(['Strength'])
    expect(testInstance.findAllByProps({ className: 'attribute' })[1].children[0].children).toEqual(['Agility'])
  })
})