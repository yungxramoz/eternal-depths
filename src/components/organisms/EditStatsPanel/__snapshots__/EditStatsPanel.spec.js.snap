// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditStatsPanel renders the correct number of attributes 1`] = `
<div
  className="rpgui-container container framed framed-golden attribute-container"
  style={Object {}}
>
  <h2>
    Attributes
  </h2>
  <p>
    Available points: 
    10
  </p>
  <hr
    className="separator"
  />
  <div
    className="attribute"
  >
    <p>
      Strength
    </p>
    <div
      className="attr-controller"
    >
      <button
        className="icon-button medium"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          id="Minus"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="2"
            width="22"
            x="1"
            y="11"
          />
        </svg>
      </button>
      <p>
        0
      </p>
      <button
        className="icon-button medium"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          id="Plus"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="attribute"
  >
    <p>
      Agility
    </p>
    <div
      className="attr-controller"
    >
      <button
        className="icon-button medium"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          id="Minus"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="2"
            width="22"
            x="1"
            y="11"
          />
        </svg>
      </button>
      <p>
        0
      </p>
      <button
        className="icon-button medium"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          id="Plus"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
